### ISSUE TITLE
Conform C++ types that have a `std::hash` overload to `Swift.Hashable`

### ISSUE THREAD

**Issue by egorzhdan** [swiftlang/swift#62045]
We could consider synthesizing a `Hashable` conformance for some C++ types:

```swift
extension MyCppType : Hashable {
  func hash(into hasher: inout Hasher) {
    hasher.combine(std.hash<MyCppType>(self))
  }
}
```

Note that `Swift.Hashable` inherits from `Swift.Equatable`, so we'd have to conform these types to `Swift.Equatable` as well.

**Comment by vmk8826** [comment]
Can you assign this to me?

**Comment by egorzhdan** [comment]
@vmk8826 sure, please feel free to try to tackle this, and don't hesitate to ask questions.

**Comment by ktoso** [comment]
Radar-ified rdar://111879435

**Comment by LukeXuan** [comment]
Is this issue currently being worked on or obsolete since 2022? If not, I can give it a try as an exercise.

I have a half working implementation (only `std::hash` without `std::equal_to`), but I'm unsure about my synthesizer code. Is there a way to pretty print the synthesized function (result of `SwiftDeclSynthesizer.makeXXXFunc(...)`) in Swift?

**Comment by LukeXuan** [comment]
Hello, I created a draft pull request so that you can check the current status. I have one question that I cannot figure out myself: How does the `==` operator find the function it invokes?

In my current [implementation](https://github.com/LukeXuan/swift/blob/2990bda3514a3bca2754bc71f351ebce21b2071b/lib/ClangImporter/ClangDerivedConformances.cpp#L1428), I had to add `@_implements(Equatable, ==(_:_:))` **after** the synthesized `==(_:_:)` method is added to the member lookup table, in order to make expressions like `a == b` to invoke this synthesized version. I tried two other approaches that I believe would have the same effect but none of them worked (`__derived_struct_equals` is invoked instead):
- calling `impl.addSynthesizedProtocolAttrs(decl, {KnownProtocolKind::Equatable});`,
- add the attribute **before** adding to member lookup table,

I'm genuinely at a loss what's happening here. Could you give me some hints on how to debug the equal operator lookup procedure? Or is there a different recommended way to synthesize a function and use it for conformance?

I also want to write some tests about this new functionality, could you point me to some existing cxx-interop test cases that I reference?

**Comment by egorzhdan** [comment]
Thanks @LukeXuan for working on this!

I would expect `ClangImporter::Implementation::lookupValue` to be invoked with the `name` argument being `==` whenever the compiler is trying to lookup a `func ==` for C++ types. If that isn't happening, I would try putting a breakpoint in `ModuleDecl::lookupValue` to see when and how we are lookup up `func ==`. Usually the fastest way to catch that is to write a little bit of debugging code to narrow down the investigation, something like `if (name.isOperator()) { ... /* put an LLDB breakpoint inside */ }`. Let me know if that's not what you're seeing!

**Comment by LukeXuan** [comment]
> I would expect `ClangImporter::Implementation::lookupValue` to be invoked with the `name` argument being `==` whenever the compiler is trying to lookup a `func ==` for C++ types.

Indeed, I found out that the clang modules use a different lookup table for operators than native swift modules, and my uneducated guess is it's an optimization for lookup performance?

In any case, my draft pull request is already using the `ClangImporter`'s lookup table for `func ==`. I will clean it up and open for review soon.
